crates/data_components/src/databricks/auth.rs

```diff
@@ -0,0 +1,192 @@
+use std::time::Duration;
+use std::{fmt, sync::Arc};
+use tokio::{sync::watch, task::JoinHandle, time::sleep};
+use util::fibonacci_backoff::FibonacciBackoffBuilder;
+
+use crate::token_provider::{Result, TokenProvider};
+
+const TOKEN_REFRESH_BUFFER_SECS: u64 = 300;
+
+#[derive(Debug, Snafu)]
+pub enum Error {
+    #[snafu(display(
+        "Failed to obtain Databricks service principal token for machine-to-machine authentication."
+    ))]
+    UnableToGetToken { source: Box<dyn std::error::Error> },
+}
+
+#[derive(Clone)]
+pub struct DatabricksM2MTokenProvider {
+    endpoint: String,
+    client_id: String,
+
+    tx: watch::Sender<String>,
+    rx: watch::Receiver<String>,
+
+    _handle: Arc<JoinHandle<()>>,
+}
+
+impl fmt::Debug for DatabricksM2MTokenProvider {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("M2mTokenProvider")
+            .field("endpoint", &self.endpoint)
+            .field("client_id", &self.client_id)
+            .field("tx", &"<watch::Sender>")
+            .field("rx", &"<watch::Receiver>")
+            .field("_handle", &"<JoinHandle>")
+            .finish()
+    }
+}
+
+impl DatabricksM2MTokenProvider {
+    pub async fn try_new(
+        endpoint: String,
+        client_id: String,
+        client_secret: SecretString,
+    ) -> Result<Self, Error> {
+        // initial fetch
+        let TokenResponse {
+            access_token,
+            expires_in,
+            ..
+        } = get_m2m_access_token(endpoint.clone(), client_id.clone(), client_secret.clone())
+            .await
+            .map_err(|e| Error::UnableToGetToken { source: e })?;
+
+        // create watch channel
+        let (tx, rx) = watch::channel(access_token.clone());
+
+        // spawn background refresh loop
+        let cloned_client_id = client_id.clone();
+        let cloned_endpoint = endpoint.clone();
+        let cloned_tx = tx.clone();
+
+        let secret = client_secret.clone();
+
+        let handle = tokio::spawn(async move {
+            // Databricks M2M access token lifespan is one hour. Schedule a refresh five minutes before expiration
+            let mut next_wait = Duration::from_secs(expires_in - TOKEN_REFRESH_BUFFER_SECS);
+
+            let mut backoff = FibonacciBackoffBuilder::new()
+                .max_duration(Some(Duration::from_secs(300))) // Cap at 5 minutes
+                .build();
+
+            loop {
+                sleep(next_wait).await;
+
+                match get_m2m_access_token(
+                    cloned_endpoint.clone(),
+                    cloned_client_id.clone(),
+                    secret.clone(),
+                )
+                .await
+                {
+                    Ok(TokenResponse {
+                        access_token,
+                        expires_in,
+                        ..
+                    }) => {
+                        tracing::info!("M2M token refreshed; expires in {}", expires_in);
```

Comment:
Maybe bump this down to a debug log

Reply:
I think we should also make this a metric

Reply:
Will do in a separate PR

Reply:
updated to use debug